### PR TITLE
[DTensor]: propagate grad_dtype through from_local and to_local

### PR DESCRIPTION
**Pull request by sorenzzio**
## Description

This PR builds on the new grad_dtype API introduced in #189633 and reimplements grad_dtype propagation across DTensor's to_local() and from_local() conversions.  For additional background, see #173990 and #174007.

**Key Changes:**
Propagated grad_dtype in DTensor operations: Updated to_local and from_local. 

**Example:**
Here is an example demonstrating this in an FSDP + EP scenario:
**AS IS:**
```text
W(dtensor) ---> (to_local) ---> W(tensor) -------> (gmm) --------> Y
dtype=bf16                      dtype=bf16                    dtype=bf16
grad_dtype=fp32

(leaf node)                     (non-leaf)                       
                                                                   
dW(dtensor) <--(to_local_bwd)-- dW(tensor) <------(gmmbwd) <------ dY
dtype=fp32                      dtype=bf16      with fp32 out   dtype=bf16
      ^                               ^                         
      |____[upcast to fp32]          |____[downcast to bf16] (align to W dtype)
```
**TO BE:**
Since grad_dtype is passed through to_local, backward gradients are no longer downcasted
```text
W(dtensor) ---> (to_local) ---> W(tensor) -------> (gmm) --------> Y
dtype=bf16                      dtype=bf16                    dtype=bf16
grad_dtype=fp32                 grad_dtype=fp32

(leaf node)                     (non-leaf)                       
                                                                   
dW(dtensor) <--(to_local_bwd)-- dW(tensor) <------(gmmbwd) <------ dY
dtype=fp32                      dtype=fp32      with fp32 out   dtype=bf16
                                      ^                         
                                      |____[no cast] (align to W grad_dtype)
```

> **AI disclosure:** The quoted PR title and description below were drafted
> with Codex, an AI coding assistant, and reviewed by the author.
>
> ## Summary
>
> The [`grad_dtype` RFC](https://github.com/pytorch/pytorch/issues/189633)
> introduces an autograd contract that lets gradients use a dtype independent
> of a tensor's storage dtype. DTensor's `from_local()` and `to_local()`
> conversions are autograd boundaries, so their outputs otherwise fall back to
> the storage dtype and lose a wider input `grad_dtype`.
>
> Propagate the input's effective `grad_dtype` to each differentiable conversion
> output with `ctx.set_output_grad_dtype(input.grad_dtype)`. The propagation is
> guarded by `input.requires_grad`, leaving non-differentiable inputs unchanged.
>
> Add a round-trip regression test using BF16 storage and FP32 gradients. The
> test checks both public `grad_dtype` values and verifies that an FP32 gradient
> value that is not exactly representable in BF16 reaches the original local
> tensor without truncation.
>
> ## Test plan
>
> Built a CPU and distributed configuration on the remote compilation host:
>
> ```bash
> USE_CUDA=0 USE_ROCM=0 USE_XPU=0 USE_NUMA=0 USE_DISTRIBUTED=1 USE_NCCL=0 USE_MPI=0 USE_TENSORPIPE=0 USE_FBGEMM=0 MAX_JOBS=64 python -m pip install -e . -v --no-build-isolation
> ```
>
> Ran the new DTensor test on four ranks:
>
> ```bash
> python test/distributed/tensor/test_dtensor.py DTensorTest.test_from_local_to_local_grad_dtype
> ```
>
> Ran the existing local-conversion regression coverage:
>
> ```bash
> python test/distributed/tensor/test_dtensor.py DTensorTest.test_from_local DTensorTest.test_from_local_backward DTensorTest.test_to_local DTensorTest.test_to_local_grad_hint DTensorTest.test_to_local_from_local_backward DTensorTest.test_from_local_then_to_local
> ```
>
> Ran the generated LocalTensor variant:
>
> ```bash
> python test/distributed/tensor/test_dtensor.py DTensorTestWithLocalTensor.test_from_local_to_local_grad_dtype
> ```
>
> Ran the integer-input regression path on eight ranks to verify the
> `requires_grad` guard:
>
> ```bash
> python test/distributed/tensor/test_dtensor.py DTensorMeshTest.test_vmap_embedding
> ```

### CODE REVIEW

**Comment by sorenzzio** [comment]
@wconstab This PR is a rework of #174007 using the newly introduced grad dtype API. Could you please review it when you have a chance?
